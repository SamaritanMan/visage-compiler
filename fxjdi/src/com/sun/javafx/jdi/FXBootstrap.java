/*
 * Copyright 2010 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package com.sun.javafx.jdi;

import com.sun.javafx.jdi.connect.FXConnector;
import com.sun.jdi.Bootstrap;
import com.sun.jdi.JDIPermission;
import com.sun.jdi.VirtualMachine;
import com.sun.jdi.VirtualMachineManager;
import com.sun.jdi.connect.AttachingConnector;
import com.sun.jdi.connect.Connector;
import com.sun.jdi.connect.LaunchingConnector;
import com.sun.jdi.connect.ListeningConnector;
import com.sun.jdi.connect.spi.Connection;
import com.sun.jdi.connect.spi.TransportService;
import com.sun.tools.jdi.GenericAttachingConnector;
import com.sun.tools.jdi.GenericListeningConnector;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.ServiceLoader;

/**
 * A simple factory to create FX-JDI conns (similar to platform provided
 * com.sun.jdi.Bootstrap - except that the default is changed here).
 * 
 * @author sundar
 */
public class FXBootstrap {

    public static VirtualMachineManager virtualMachineManager() {
        return FXVirtualMachineManager.virtualMachineManager();
    }

    private static class FXVirtualMachineManager implements VirtualMachineManager {
        private List<Connector> connectors = new ArrayList<Connector>();
        private LaunchingConnector defaultConnector = null;
        private static final int majorVersion = 1;
        private static final int minorVersion = 6;
        private static final Object lock = new Object();
        private static FXVirtualMachineManager vmm;

        public static VirtualMachineManager virtualMachineManager() {
            SecurityManager sm = System.getSecurityManager();
            if (sm != null) {
                JDIPermission vmmPermission =
                        new JDIPermission("virtualMachineManager");
                sm.checkPermission(vmmPermission);
            }
            synchronized (lock) {
                if (vmm == null) {
                    vmm = new FXVirtualMachineManager();
                }
            }
            return vmm;
        }

        protected FXVirtualMachineManager() {
            /*
             * Load the connectors
             */
            ServiceLoader<Connector> connectorLoader =
                    ServiceLoader.load(Connector.class, FXConnector.class.getClassLoader());

            Iterator<Connector> conns = connectorLoader.iterator();

            while (conns.hasNext()) {
                Connector connector;

                try {
                    connector = conns.next();
                } catch (ThreadDeath ex) {
                    throw ex;
                } catch (Exception ex) {
                    System.err.println(ex);
                    continue;
                } catch (Error ex) {
                    System.err.println(ex);
                    continue;
                }

                addConnector(connector);
            }

            /*
             * Load any transport services and encapsulate them with
             * an attaching and listening connector.
             */
            ServiceLoader<TransportService> transportLoader =
                    ServiceLoader.load(TransportService.class,
                    TransportService.class.getClassLoader());

            Iterator<TransportService> transportServices =
                    transportLoader.iterator();

            while (transportServices.hasNext()) {
                TransportService transportService;

                try {
                    transportService = transportServices.next();
                } catch (ThreadDeath ex) {
                    throw ex;
                } catch (Exception ex) {
                    System.err.println(ex);
                    continue;
                } catch (Error ex) {
                    System.err.println(ex);
                    continue;
                }

                addConnector(GenericAttachingConnector.create(transportService));
                addConnector(GenericListeningConnector.create(transportService));
            }

            // no connectors found
            if (allConnectors().size() == 0) {
                throw new Error("no Connectors loaded");
            }

            // default is FX-JDI launching connector
            boolean found = false;
            List<LaunchingConnector> launchers = launchingConnectors();
            for (LaunchingConnector lc : launchers) {
                if (lc.name().equals("com.sun.javafx.jdi.connect.FXLaunchingConnector")) {
                    setDefaultConnector(lc);
                    found = true;
                    break;
                }
            }
            if (!found && launchers.size() > 0) {
                setDefaultConnector(launchers.get(0));
            }

        }

        public LaunchingConnector defaultConnector() {
            if (defaultConnector == null) {
                throw new Error("no default LaunchingConnector");
            }
            return defaultConnector;
        }

        public void setDefaultConnector(LaunchingConnector connector) {
            defaultConnector = connector;
        }

        public List<LaunchingConnector> launchingConnectors() {
            List<LaunchingConnector> launchingConnectors = new ArrayList<LaunchingConnector>(connectors.size());
            for (Connector connector : connectors) {
                if (connector instanceof LaunchingConnector) {
                    launchingConnectors.add((LaunchingConnector) connector);
                }
            }
            return Collections.unmodifiableList(launchingConnectors);
        }

        public List<AttachingConnector> attachingConnectors() {
            List<AttachingConnector> attachingConnectors = new ArrayList<AttachingConnector>(connectors.size());
            for (Connector connector : connectors) {
                if (connector instanceof AttachingConnector) {
                    attachingConnectors.add((AttachingConnector) connector);
                }
            }
            return Collections.unmodifiableList(attachingConnectors);
        }

        public List<ListeningConnector> listeningConnectors() {
            List<ListeningConnector> listeningConnectors = new ArrayList<ListeningConnector>(connectors.size());
            for (Connector connector : connectors) {
                if (connector instanceof ListeningConnector) {
                    listeningConnectors.add((ListeningConnector) connector);
                }
            }
            return Collections.unmodifiableList(listeningConnectors);
        }

        public List<Connector> allConnectors() {
            return Collections.unmodifiableList(connectors);
        }

        public List<VirtualMachine> connectedVirtualMachines() {
            VirtualMachineManager pvmm = Bootstrap.virtualMachineManager();
            return FXWrapper.wrapVirtualMachines(pvmm.connectedVirtualMachines());
        }

        public void addConnector(Connector connector) {
            connectors.add(connector);
        }

        public void removeConnector(Connector connector) {
            connectors.remove(connector);
        }

        public VirtualMachine createVirtualMachine(Connection connection, 
                Process process) throws IOException {
            VirtualMachineManager pvmm = Bootstrap.virtualMachineManager();
            return FXWrapper.wrap(pvmm.createVirtualMachine(connection, process));
        }

        public VirtualMachine createVirtualMachine(Connection connection) 
                throws IOException {
            VirtualMachineManager pvmm = Bootstrap.virtualMachineManager();
            return FXWrapper.wrap(pvmm.createVirtualMachine(connection));
        }

        public int majorInterfaceVersion() {
            return majorVersion;
        }

        public int minorInterfaceVersion() {
            return minorVersion;
        }
    }
}
