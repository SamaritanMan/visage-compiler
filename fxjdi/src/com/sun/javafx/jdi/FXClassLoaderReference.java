/*
 * Copyright 2010 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package com.sun.javafx.jdi;

import com.sun.jdi.ClassLoaderReference;
import com.sun.jdi.ReferenceType;
import java.util.List;

/**
 *
 * @author sundar
 */
public class FXClassLoaderReference extends FXObjectReference implements ClassLoaderReference {
    public FXClassLoaderReference(FXVirtualMachine fxvm, ClassLoaderReference underlying) {
        super(fxvm, underlying);
    }

    public List<ReferenceType> definedClasses() {
        return FXWrapper.wrapReferenceTypes(virtualMachine(), underlying().definedClasses());
    }

    public List<ReferenceType> visibleClasses() {
        return FXWrapper.wrapReferenceTypes(virtualMachine(), underlying().visibleClasses());
    }

    @Override
    protected ClassLoaderReference underlying() {
        return (ClassLoaderReference) super.underlying();
    }
}
