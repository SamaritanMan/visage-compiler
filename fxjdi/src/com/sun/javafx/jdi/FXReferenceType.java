/*
 * Copyright 2010 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package com.sun.javafx.jdi;

import com.sun.jdi.AbsentInformationException;
import com.sun.jdi.ClassLoaderReference;
import com.sun.jdi.ClassObjectReference;
import com.sun.jdi.Field;
import com.sun.jdi.Location;
import com.sun.jdi.Method;
import com.sun.jdi.ObjectReference;
import com.sun.jdi.ReferenceType;
import com.sun.jdi.Value;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 *
 * @author sundar
 */
public class FXReferenceType extends FXType implements ReferenceType {
    public FXReferenceType(FXVirtualMachine fxvm, ReferenceType underlying) {
        super(fxvm, underlying);
    }

    public byte[] constantPool() {
        return underlying().constantPool();
    }

    public int constantPoolCount() {
        return underlying().constantPoolCount();
    }

    public List<ObjectReference> instances(long count) {
        return FXWrapper.wrapObjectReferences(virtualMachine(), underlying().instances(count));
    }

    public int majorVersion() {
        return underlying().majorVersion();
    }

    public int minorVersion() {
        return underlying().minorVersion();
    }

    public List<Field> allFields() {
        return FXWrapper.wrapFields(virtualMachine(), underlying().allFields());
    }

    public List<Location> allLineLocations() throws AbsentInformationException {
        return FXWrapper.wrapLocations(virtualMachine(), underlying().allLineLocations());
    }

    public List<Location> allLineLocations(String stratum, String sourceName) throws AbsentInformationException {
        return FXWrapper.wrapLocations(virtualMachine(), underlying().allLineLocations(stratum, sourceName));
    }

    public List<Method> allMethods() {
        return FXWrapper.wrapMethods(virtualMachine(), underlying().allMethods());
    }

    public List<String> availableStrata() {
        return underlying().availableStrata();
    }

    public ClassLoaderReference classLoader() {
        return FXWrapper.wrap(virtualMachine(), underlying().classLoader());
    }

    public ClassObjectReference classObject() {
        return FXWrapper.wrap(virtualMachine(), underlying().classObject());
    }

    public String defaultStratum() {
        return underlying().defaultStratum();
    }

    public boolean failedToInitialize() {
        return underlying().failedToInitialize();
    }

    // return null if there is no field or the name is ambigous. 
    public FXField fieldByName(String name) {
        // There could be both an FX field $xxx and a java field xxx
        Field javaField = underlying().fieldByName(name);
        Field fxField = underlying().fieldByName("$" + name);
        if (javaField == null) {
            if (fxField == null ) {
                return null;
            }
            // we'll return fxField
        } else {
            if (fxField != null) {
                // we found both name and $name
                return null;
            }
            fxField = javaField;
        }
        return FXWrapper.wrap(virtualMachine(), fxField);
    }

    public List<Field> fields() {
        return FXWrapper.wrapFields(virtualMachine(), underlying().fields());
    }

    public String genericSignature() {
        return underlying().genericSignature();
    }

    public FXValue getValue(Field field) {
        if (!FXUtils.isValid(this, field)) {
            throw new FXInvalidValueException(this, field);
        }
        return FXWrapper.wrap(virtualMachine(), underlying().getValue(FXWrapper.unwrap(field)));
    }

    public Map<Field, Value> getValues(List<? extends Field> fields) {
        Map<Field, Field> fieldMap = new HashMap<Field, Field>();
        List<Field> unwrappedFields = new ArrayList<Field>();
        for (Field field : fields) {
            Field unwrapped = FXWrapper.unwrap(field);
            unwrappedFields.add(unwrapped);
            fieldMap.put(unwrapped, field);
        }
        Map<Field, Value> fieldValues = underlying().getValues(unwrappedFields);
        Map<Field, Value> result = new HashMap<Field, Value>();
        for (Map.Entry<Field, Value> entry: fieldValues.entrySet()) {
            result.put(fieldMap.get(entry.getKey()), entry.getValue());
        }
        return result;
    }

    public boolean isAbstract() {
        return underlying().isAbstract();
    }

    public boolean isFinal() {
        return underlying().isFinal();
    }

    public boolean isInitialized() {
        return underlying().isInitialized();
    }

    public boolean isPrepared() {
        return underlying().isPrepared();
    }

    public boolean isStatic() {
        return underlying().isStatic();
    }

    public boolean isVerified() {
        return underlying().isVerified();
    }

    public List<Location> locationsOfLine(int lineNumber) throws AbsentInformationException {
        return FXWrapper.wrapLocations(virtualMachine(), underlying().locationsOfLine(lineNumber));
    }

    public List<Location> locationsOfLine(String stratum, String sourceName, int line) throws AbsentInformationException {
        return FXWrapper.wrapLocations(virtualMachine(), underlying().locationsOfLine(stratum, sourceName, line));
    }

    public List<Method> methods() {
        return FXWrapper.wrapMethods(virtualMachine(), underlying().methods());
    }

    public List<Method> methodsByName(String name) {
        return FXWrapper.wrapMethods(virtualMachine(), underlying().methodsByName(name));
    }

    public List<Method> methodsByName(String name, String signature) {
        return FXWrapper.wrapMethods(virtualMachine(), underlying().methodsByName(name, signature));
    }

    public List<ReferenceType> nestedTypes() {
        return FXWrapper.wrapReferenceTypes(virtualMachine(), underlying().nestedTypes());
    }

    public String sourceDebugExtension() throws AbsentInformationException {
        return underlying().sourceDebugExtension();
    }

    public String sourceName() throws AbsentInformationException {
        return underlying().sourceName();
    }

    public List<String> sourceNames(String stratum) throws AbsentInformationException {
        return underlying().sourceNames(stratum);
    }

    public List<String> sourcePaths(String stratum) throws AbsentInformationException {
        return underlying().sourcePaths(stratum);
    }

    public List<Field> visibleFields() {
        return FXWrapper.wrapFields(virtualMachine(), underlying().visibleFields());
    }

    public List<Method> visibleMethods() {
        return FXWrapper.wrapMethods(virtualMachine(), underlying().visibleMethods());
    }

    public int compareTo(ReferenceType o) {
        return underlying().compareTo(FXWrapper.unwrap(o));
    }

    public boolean isPackagePrivate() {
        return underlying().isPackagePrivate();
    }

    public boolean isPrivate() {
        return underlying().isPrivate();
    }

    public boolean isProtected() {
        return underlying().isProtected();
    }

    public boolean isPublic() {
        return underlying().isPublic();
    }

    public int modifiers() {
        return underlying().modifiers();
    }

    @Override
    protected ReferenceType underlying() {
        return (ReferenceType) super.underlying();
    }

    protected boolean isJavaFXType() {
        return false;
    }
}
