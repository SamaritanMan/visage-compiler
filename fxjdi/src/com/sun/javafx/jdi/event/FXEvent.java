/*
 * Copyright 2010 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package com.sun.javafx.jdi.event;

import com.sun.javafx.jdi.FXMirror;
import com.sun.javafx.jdi.FXVirtualMachine;
import com.sun.javafx.jdi.request.FXEventRequest;
import com.sun.jdi.event.AccessWatchpointEvent;
import com.sun.jdi.event.BreakpointEvent;
import com.sun.jdi.event.ClassPrepareEvent;
import com.sun.jdi.event.ClassUnloadEvent;
import com.sun.jdi.event.Event;
import com.sun.jdi.event.ExceptionEvent;
import com.sun.jdi.event.LocatableEvent;
import com.sun.jdi.event.MethodEntryEvent;
import com.sun.jdi.event.MethodExitEvent;
import com.sun.jdi.event.ModificationWatchpointEvent;
import com.sun.jdi.event.MonitorContendedEnterEvent;
import com.sun.jdi.event.MonitorContendedEnteredEvent;
import com.sun.jdi.event.MonitorWaitEvent;
import com.sun.jdi.event.MonitorWaitedEvent;
import com.sun.jdi.event.StepEvent;
import com.sun.jdi.event.ThreadDeathEvent;
import com.sun.jdi.event.ThreadStartEvent;
import com.sun.jdi.event.VMDeathEvent;
import com.sun.jdi.event.VMDisconnectEvent;
import com.sun.jdi.event.VMStartEvent;
import com.sun.jdi.event.WatchpointEvent;
import java.util.ArrayList;
import java.util.Collection;

/**
 *
 * @author sundar
 */
public class FXEvent extends FXMirror implements Event {
    public FXEvent(FXVirtualMachine fxvm, Event underlying) {
        super(fxvm, underlying);
    }

    public FXEventRequest request() {
        return FXEventRequest.wrap(virtualMachine(), underlying().request());
    }

    @Override
    protected Event underlying() {
        return (Event) super.underlying();
    }

    public static Event unwrap(Event evt) {
        return (evt instanceof FXEvent)? ((FXEvent)evt).underlying() : evt;
    }

    public static FXEvent wrap(FXVirtualMachine fxvm, Event evt) {
        if (evt == null) {
            return null;
        }

        if (evt instanceof AccessWatchpointEvent) {
            return new FXAccessWatchpointEvent(fxvm, (AccessWatchpointEvent)evt);
        } else if (evt instanceof BreakpointEvent) {
            return new FXBreakpointEvent(fxvm, (BreakpointEvent)evt);
        } else if (evt instanceof ClassPrepareEvent) {
            return new FXClassPrepareEvent(fxvm, (ClassPrepareEvent)evt);
        } else if (evt instanceof ClassUnloadEvent) {
            return new FXClassUnloadEvent(fxvm, (ClassUnloadEvent)evt);
        } else if (evt instanceof ExceptionEvent) {
            return new FXExceptionEvent(fxvm, (ExceptionEvent)evt);
        } else if (evt instanceof MethodEntryEvent) {
            return new FXMethodEntryEvent(fxvm, (MethodEntryEvent)evt);
        } else if (evt instanceof MethodExitEvent) {
            return new FXMethodExitEvent(fxvm, (MethodExitEvent)evt);
        } else if (evt instanceof ModificationWatchpointEvent) {
            return new FXModificationWatchpointEvent(fxvm, (ModificationWatchpointEvent)evt);
        } else if (evt instanceof MonitorContendedEnterEvent) {
            return new FXMonitorContendedEnterEvent(fxvm, (MonitorContendedEnterEvent)evt);
        } else if (evt instanceof MonitorContendedEnteredEvent) {
            return new FXMonitorContendedEnteredEvent(fxvm, (MonitorContendedEnteredEvent)evt);
        } else if (evt instanceof MonitorWaitEvent) {
            return new FXMonitorWaitEvent(fxvm, (MonitorWaitEvent)evt);
        } else if (evt instanceof MonitorWaitedEvent) {
            return new FXMonitorWaitedEvent(fxvm, (MonitorWaitedEvent)evt);
        } else if (evt instanceof StepEvent) {
            return new FXStepEvent(fxvm, (StepEvent)evt);
        } else if (evt instanceof ThreadDeathEvent) {
            return new FXThreadDeathEvent(fxvm, (ThreadDeathEvent)evt);
        } else if (evt instanceof ThreadStartEvent) {
            return new FXThreadStartEvent(fxvm, (ThreadStartEvent)evt);
        } else if (evt instanceof VMDeathEvent) {
            return new FXVMDeathEvent(fxvm, (VMDeathEvent)evt);
        } else if (evt instanceof VMDisconnectEvent) {
            return new FXVMDisconnectEvent(fxvm, (VMDisconnectEvent)evt);
        } else if (evt instanceof VMStartEvent) {
            return new FXVMStartEvent(fxvm, (VMStartEvent)evt);
        } else if (evt instanceof WatchpointEvent) {
            return new FXWatchpointEvent(fxvm, (WatchpointEvent)evt);
        } else if (evt instanceof LocatableEvent) {
            return new FXLocatableEvent(fxvm, (LocatableEvent)evt);
        } else {
            return new FXEvent(fxvm, evt);
        }
    }

    public static Collection unwrapEvents(Collection events) {
        if (events == null) {
            return null;
        }
        ArrayList result = new ArrayList();
        for (Object obj : events) {
            result.add((obj instanceof Event)? unwrap((Event)obj) : obj);
        }
        return result;
    }
}