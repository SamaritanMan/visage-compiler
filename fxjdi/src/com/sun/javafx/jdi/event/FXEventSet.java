/*
 * Copyright 2010 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

package com.sun.javafx.jdi.event;

import com.sun.javafx.jdi.FXMirror;
import com.sun.javafx.jdi.FXVirtualMachine;
import com.sun.jdi.event.Event;
import com.sun.jdi.event.EventSet;
import java.util.Collection;
import java.util.Iterator;

/**
 *
 * @author sundar
 */
public class FXEventSet extends FXMirror implements EventSet {
    public FXEventSet(FXVirtualMachine fxvm, EventSet underlying) {
        super(fxvm, underlying);
    }

    public FXEventIterator eventIterator() {
        return FXEventIterator.wrap(virtualMachine(), underlying().eventIterator());
    }

    public void resume() {
        underlying().resume();
    }

    public int suspendPolicy() {
        return underlying().suspendPolicy();
    }

    public boolean add(Event evt) {
        return underlying().add(FXEvent.unwrap(evt));
    }

    public boolean addAll(Collection<? extends Event> events) {
        return underlying().addAll(FXEvent.unwrapEvents(events));
    }

    public void clear() {
        underlying().clear();
    }

    public boolean contains(Object obj) {
        return underlying().contains((obj instanceof Event)? FXEvent.unwrap((Event)obj) : obj);
    }

    public boolean containsAll(Collection<?> arg0) {
        return underlying().containsAll(FXEvent.unwrapEvents(arg0));
    }

    public boolean isEmpty() {
        return underlying().isEmpty();
    }

    public Iterator<Event> iterator() {
        final Iterator<Event> wrapped = underlying().iterator();
        if (wrapped != null) {
            return new Iterator<Event>() {
                public boolean hasNext() {
                    return wrapped.hasNext();
                }

                public Event next() {
                    return FXEvent.wrap(virtualMachine(), wrapped.next());
                }

                public void remove() {
                    wrapped.remove();
                }
            };
        } else {
            return null;
        }
    }

    public boolean remove(Object obj) {
        return underlying().remove((obj instanceof Event)?
            FXEvent.unwrap((Event)obj) : obj);
    }

    public boolean removeAll(Collection<?> arg0) {
        return underlying().removeAll(FXEvent.unwrapEvents(arg0));
    }

    public boolean retainAll(Collection<?> arg0) {
        return underlying().retainAll(FXEvent.unwrapEvents(arg0));
    }

    public int size() {
        return underlying().size();
    }

    public Object[] toArray() {
        Object[] res = underlying().toArray();
        if (res != null) {
            for (int i = 0; i < res.length; i++) {
                if (res[i] instanceof Event) {
                    res[i] = FXEvent.unwrap((Event)res[i]);
                }
            }
        }
        return res;
    }

    public <T> T[] toArray(T[] arg0) {
        // FIXME
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    protected EventSet underlying() {
        return (EventSet) super.underlying();
    }

    public static FXEventSet wrap(FXVirtualMachine fxvm, EventSet evtSet) {
        return (evtSet == null)? null : new FXEventSet(fxvm, evtSet);
    }
}