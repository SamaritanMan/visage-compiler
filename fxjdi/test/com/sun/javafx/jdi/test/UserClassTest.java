/*
 * Copyright 2010 Sun Microsystems, Inc.  All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
 * CA 95054 USA or visit www.sun.com if you need additional information or
 * have any questions.
 */

/*
 * This is a basic test to show that internal instance vars are properly filtered out
 * and that the internal naming scheme for user instance vars is filtered out.
 */

package com.sun.javafx.jdi.test;

import com.sun.jdi.ReferenceType;
import com.sun.jdi.ThreadReference;
import com.sun.jdi.StackFrame;
import com.sun.jdi.ClassType;
import com.sun.javafx.jdi.FXReferenceType;
import com.sun.jdi.event.BreakpointEvent;
import java.util.List;
import org.junit.Test;
import junit.framework.Assert;

public class UserClassTest extends JavafxTestBase {
    ReferenceType targetClass;
    ThreadReference mainThread;

    private static final String targetClassName = "com.sun.javafx.jdi.test.target.UserClassTarget";

    public UserClassTest() {
        super(targetClassName);
    }

    @Test
    public void testUserClass() {
        try {
            startTests();
        } catch (Exception exp) {
            exp.printStackTrace();
            Assert.fail(exp.getMessage());
        }
    }

    /********** test core **********/

    void printUser(StackFrame frame, int count) {
        ReferenceType rt = frame.location().declaringType();
        ClassType userClass = ((FXReferenceType)rt).javaFXUserClass();
        writeActual("userClass " + count + " = " + rt.name() + ", " + 
                    (userClass == null? null: userClass.name()));
    }

    protected void runTests() throws Exception {
        startToMain();
        BreakpointEvent bpe = resumeTo("javafx.lang.Builtins", "println", "(Ljava/lang/Object;)V");
        mainThread = bpe.thread();

        targetClass = bpe.location().declaringType();
        // We are stopped under the on replace for sv2
            
        List<StackFrame> frames =  bpe.thread().frames(0,10);
        for (StackFrame frame: frames) {
            printUser(frame, 1);
        }

        bpe = resumeTo("javafx.lang.Builtins", "println", "(Ljava/lang/Object;)V");
        // stopped in the on replace for override1
        printUser(bpe.thread().frame(2), 2);
        testFailed = !didTestPass();

        /*
         * resume until end
         */
        listenUntilVMDisconnect();
        
        /*
         * deal with results of test
         * if anything has called failure("foo") testFailed will be true
         */
        if (!testFailed) {
            writeActual(testClassName + ": passed");
        } else {
            throw new Exception(testClassName + ": failed");
        }
    }
}
