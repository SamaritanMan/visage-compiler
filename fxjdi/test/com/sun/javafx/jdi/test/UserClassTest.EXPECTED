userClass 1 = javafx.lang.Builtins, null
userClass 1 = com.sun.javafx.jdi.test.target.UserClassTarget, null
userClass 1 = com.sun.javafx.jdi.test.target.UserClassTarget, null
userClass 1 = com.sun.javafx.jdi.test.target.UserClassTarget$UserClassTarget$Script, com.sun.javafx.jdi.test.target.UserClassTarget
userClass 1 = com.sun.javafx.jdi.test.target.UserClassTarget, null
userClass 1 = sun.reflect.NativeMethodAccessorImpl, null
userClass 1 = sun.reflect.NativeMethodAccessorImpl, null
userClass 1 = sun.reflect.DelegatingMethodAccessorImpl, null
userClass 1 = java.lang.reflect.Method, null
userClass 1 = com.sun.javafx.runtime.Entry$NoRuntimeDefault, null
userClass 2 = com.sun.javafx.jdi.test.target.UserClassTarget$1UserClassTarget$ObjLit$1, com.sun.javafx.jdi.test.target.UserClassTarget
